### ISSUE TITLE
Switch to dataclass for data flow

### ISSUE THREAD

**Issue by ATriantafyllopoulos** [autrainer/autrainer#41]
Currently, we rely on tuples for our data flow, i.e. [this is what our datasets return](https://github.com/autrainer/autrainer/blob/d1ca7efbb2a42961ebf4313ef8546e8b1856678a/autrainer/datasets/utils/dataset_wrapper.py#L88) and this is what everything else expects, e.g. [the training loop](https://github.com/autrainer/autrainer/blob/d1ca7efbb2a42961ebf4313ef8546e8b1856678a/autrainer/training/training.py#L454)

I think it's more modular to switch to `@dataclass` objects. This allows us to extend our framework more flexibly, especially regarding models. For example, I am thinking about introducing new tasks that might require additional inputs to the model that come from the dataset. I think the easiest way is to have a base `@dataclass` object that I would then override to add additional parameters that need to be carried around.

Potentially, we need to use this for model outputs too, and use it as the interface between models/losses/transforms.

**Comment by ramppdev** [comment]
I like the general idea of using one dataclass instance per returned sample of the dataset, but wouldnt this highly complicate creating batches of data, targets, etc and or moving them to GPU / through the model?

But yes i really see the benefit of requiring e.g. `features`, `target`, and `sample_idx` to be present in the dataclass, and any additional fields like a `mask` or sth could be automatically fed into the model for modularity.

**Comment by ATriantafyllopoulos** [comment]
> but wouldnt this highly complicate creating batches of data, targets, etc and or moving them to GPU / through the model?

Not necessarily, we just need to provide a generic `collate_fn`. A quick version could be:

```python
from dataclasses import dataclass, fields
import torch
import typing

@dataclass
class Data:
    foo: torch.Tensor
    bar: torch.Tensor

class Dataset(torch.utils.data.Dataset):
    def __init__(self, data):
        self.data = data
    def __len__(self):
        return len(self.data)
    def __getitem__(self, item):
        return self.data[item]

def collate_fn(batch: typing.List[Data]) -> Data:
    results = {
        k.name: torch.stack([getattr(elem, k.name) for elem in batch])
        for k in fields(batch[0])
    }
    return Data(**results)


x = Data(torch.Tensor([1]), torch.Tensor([2]))
y = Data(torch.Tensor([3]), torch.Tensor([4]))

dataset = Dataset([x, y])

loader = torch.utils.data.DataLoader(dataset, batch_size=2, collate_fn=collate_fn)
print(next(iter(loader)))
```

or we could also call the one [from huggingface](https://github.com/huggingface/transformers/blob/53fad641cfdb5105e2470bcf3ef17ea8e25cc300/src/transformers/data/data_collator.py#L127) though I don't like all aspects of it..

**Comment by ramppdev** [comment]
Good idea, but we need to be cautious about augmentations that could possibly provide collate functions, see: https://autrainer.github.io/autrainer/modules/augmentations.html#augmentations-with-collate-functions

**Comment by ATriantafyllopoulos** [comment]
> Good idea, but we need to be cautious about augmentations that could possibly provide collate functions, see: https://autrainer.github.io/autrainer/modules/augmentations.html#augmentations-with-collate-functions

Yes, I'm aware of that, they would need to override/extend the default `collate_fn`, right?

**Comment by ramppdev** [comment]
Yes, just mentioned it to keep it in mind. For CutMix etc. we probably have to destructure the dataclass and build it back up, but should not be too complicated.

But we definitely should force certain field names to always be present which could be: features (or data), target, item (or idx).
If a dataset returns some additional fields, these could be passed to the model iff it has corresponding arguments in the forward method.
Now the question is from which position we want to start using dataclasses, it would probably make sense to use them in transforms as well? I imagine a situation where the label / target should be transformed alongside the features?

**Comment by ATriantafyllopoulos** [comment]
> But we definitely should force certain field names to always be present which could be: features (or data), target, item (or idx).

Absolutely!

> If a dataset returns some additional fields, these could be passed to the model iff it has corresponding arguments in the forward method.

My idea was to pass the object itself, and let the model consume any arguments it wishes to, i.e. not check on the call-side..

> Now the question is from which position we want to start using dataclasses, it would probably make sense to use them in transforms as well? I imagine a situation where the label / target should be transformed alongside the features?

Yes, I think that's nice to have in the long term (for instance, when mixing data with noise, you might want to change the label)

**Comment by ramppdev** [comment]
Sounds like a plan, then lets use dataclasses for transforms and augmentations as well.
This will be a quite large change so could take a bit of time, so i would say we postpone it to 0.5.0 and finish up 0.4.0 first?

**Comment by ramppdev** [comment]
One thing that would influence the API and all users with custom models is in the `forward` method which is then tasked with selecting the relevant fields of the dataclass passed right?
